keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlLifecycleEventRepository.kt

```diff
@@ -30,20 +31,56 @@ class SqlLifecycleEventRepository(
 ) : LifecycleEventRepository {
   private val log by lazy { LoggerFactory.getLogger(javaClass) }
 
+  /**
+   * Only saves event if a field other than the timestamp changes.
+   * If only the timestamp changed, update the timestamp on the existing row.
+   */
   override fun saveEvent(event: LifecycleEvent) {
     val timestamp = event.timestamp?.toTimestamp() ?: clock.timestamp()
     sqlRetry.withRetry(WRITE) {
-      jooq.insertInto(LIFECYCLE_EVENT)
-        .set(LIFECYCLE_EVENT.UID, ULID().nextULID(clock.millis()))
-        .set(LIFECYCLE_EVENT.SCOPE, event.scope.name)
-        .set(LIFECYCLE_EVENT.REF, event.artifactRef)
-        .set(LIFECYCLE_EVENT.ARTIFACT_VERSION, event.artifactVersion)
-        .set(LIFECYCLE_EVENT.TYPE, event.type.name)
-        .set(LIFECYCLE_EVENT.ID, event.id)
-        .set(LIFECYCLE_EVENT.STATUS, event.status.name)
-        .set(LIFECYCLE_EVENT.TIMESTAMP, timestamp)
-        .set(LIFECYCLE_EVENT.JSON, objectMapper.writeValueAsString(event))
-        .execute()
+      jooq.transaction { config ->
+        val txn = DSL.using(config)
+        // if event exists, only update timestamp
+        var uid: String? = null
+        txn.select(LIFECYCLE_EVENT.UID, LIFECYCLE_EVENT.JSON)
+          .from(LIFECYCLE_EVENT)
+          .where(LIFECYCLE_EVENT.SCOPE.eq(event.scope.name))
+          .and(LIFECYCLE_EVENT.REF.eq(event.artifactRef))
+          .and(LIFECYCLE_EVENT.ARTIFACT_VERSION.eq(event.artifactVersion))
+          .and(LIFECYCLE_EVENT.TYPE.eq(event.type.name))
+          .and(LIFECYCLE_EVENT.ID.eq(event.id))
+          .and(LIFECYCLE_EVENT.STATUS.eq(event.status.name))
+          .limit(1)
+          .fetch()
+          .map { (savedUid: String, savedEvent: String) ->
+            uid = savedUid
+            objectMapper.readValue<LifecycleEvent>(savedEvent)
+          }
+          .firstOrNull()
+          ?.let { existingEvent: LifecycleEvent ->
+            if (existingEvent.copy(timestamp = null) == event.copy(timestamp = null)) {
```

Comment:
You could simplify this like so:
```suggestion
            if (event == existingEvent.copy(timestamp = event.timestamp)) {
```

Reply:
yes nice, that is better